### PR TITLE
Class components

### PR DESCRIPTION
**Pull request by berendkleinhaneveld**
Add a `Component` class which can be subclassed to create components with a lifecycle.

### CODE REVIEW

File: collagraph/__init__.py [hunk 1]
```diff
@@ -439,7 +516,30 @@ def is_equivalent_event_handler(val, other, key):
                 continue
 
             event_type = key_to_event(name)
-            self.renderer.add_event_listener(dom, event_type, next_props[name])
+            events_to_add[event_type] = next_props[name]
+
+        if fiber.parent and fiber.parent.component:
+            if events_to_remove or events_to_add or attrs_to_remove or attrs_to_update:
+                if fiber.parent.component._mounted:
+                    fiber.parent.component.before_update()
```
Comment (berendkleinhaneveld) [inline collagraph/__init__.py:524]: This feels very brittle. I don't have a better solution yet though.

File: collagraph/__init__.py [hunk 2]
```diff
@@ -439,7 +516,30 @@ def is_equivalent_event_handler(val, other, key):
...
+            for key, val in attrs_to_update.items():
+                self.renderer.set_attribute(dom, key, val)
+
+            for event_type, val in events_to_add.items():
+                self.renderer.add_event_listener(dom, event_type, val)
+
+        if fiber.parent and fiber.parent.component:
+            if events_to_remove or events_to_add or attrs_to_remove or attrs_to_update:
+                if fiber.parent.component._mounted:
+                    fiber.parent.component.updated()
```
Comment (berendkleinhaneveld) [inline collagraph/__init__.py:542]: Same here, this needs an overhaul.

File: collagraph/__init__.py [hunk 3]
```diff
@@ -156,7 +165,11 @@ def work_loop(self, deadline: int):
     def perform_unit_of_work(self, fiber: Fiber) -> Optional[Fiber]:
         is_function_component = callable(fiber.type)
         if is_function_component:
```
Comment (berendkleinhaneveld) [inline collagraph/__init__.py:167]: This 'switch' statement can be improved.

File: examples/component-example.py [hunk 4]
```diff
@@ -0,0 +1,122 @@
...
+
+    controls = gfx.OrbitControls(camera.position.clone())
+    controls.add_default_event_handlers(renderer, canvas, camera)
+
+    gui = Collagraph(renderer=PygfxRenderer(), event_loop_type=EventLoopType.QT)
+
+    # Should be possible to create this element
+    # by rendering JSX to dict.
+    # Jinja supports rendering templates to Python objects
+    # so could be an interesting starting place.
```
Comment (berendkleinhaneveld) [inline examples/component-example.py:109]:
```suggestion
```

File: examples/component-example.py [hunk 5]
```diff
@@ -0,0 +1,122 @@
...
+
+    def mouse_up(self, event):
+        self._captured = False
+
+    def render(self):
+        return h(
+            "Mesh",
+            {
+                **self.props,
+                **self.state,
```
Comment (berendkleinhaneveld) [inline examples/component-example.py:69]: Only put the props/state that are actually applied here.

File: collagraph/__init__.py [hunk 6]
```diff
@@ -170,6 +183,31 @@ def perform_unit_of_work(self, fiber: Fiber) -> Optional[Fiber]:
...
+            component = fiber.type(fiber.props)
+
+        # Attach the component instance to the fiber
+        fiber.component = component
+        fiber.component_watcher = watch(
+            # FIXME: sometimes fiber.component is None
+            # Seems to happen after a while after generating lots of changes.
+            # Could be that it is only when updates are coming in faster than
+            # they are rendered...
+            lambda: fiber.component and fiber.component.state,
```
Comment (berendkleinhaneveld) [inline collagraph/__init__.py:198]: Might also need to watch `fiber.component.props`?

File: collagraph/component.py [hunk 7]
```diff
@@ -0,0 +1,79 @@
...
+
+    def __init__(self, props=None):
+        self._props = readonly(props)
+        self._state = reactive({})
+        self._mounted = False
+
+    @property
+    def state(self):
+        """Readonly view on local component state."""
+        return readonly(self._state)
```
Comment (Korijn) [inline collagraph/component.py:17]: I don't really follow why this is readonly, and why you don't just store state and props on public attributes on self?

File: collagraph/component.py [hunk 8]
```diff
@@ -0,0 +1,79 @@
...
+from observ import reactive, readonly
+
+
+class Component(metaclass=ABCMeta):
+    """Abstract base class for components"""
+
+    def __init__(self, props=None):
+        self._props = readonly(props)
+        self._state = reactive({})
+        self._mounted = False
```
Comment (Korijn) [inline collagraph/component.py:12]: I don't really follow why this `self._props` is private and readonly, and why `self._state` is private and why `self.state` is readonly?

File: examples/component-example.py [hunk 9]
```diff
@@ -0,0 +1,122 @@
...
+class Scrubber(Component):
+    scrubber_geometry = gfx.sphere_geometry()
+    scrubber_material = gfx.MeshPhongMaterial(color=[0.5, 1, 0.5])
+
+    def __init__(self, props):
+        super().__init__(props)
+        self._captured = False
+        self._sphere = None
+        self._mouse_pos = (0, 0)
+        self.position = [0, 0, 0]
```
Comment (Korijn) [inline examples/component-example.py:32]: so, if I understand the `__setattr__` and `__getattribute__` overrides in `Component` correctly, `self.position` is state, but the private fields are not? unless you instantiate `Scrubber` with a `position` prop, then it's not state, but a prop. Right?

File: examples/component-example.py [hunk 10]
```diff
@@ -0,0 +1,122 @@
...
+
+    def mouse_down(self, event):
+        if not self._sphere:
+            self._sphere = event.current_target
```
Comment (Korijn) [inline examples/component-example.py:36]: the challenge here is getting a reference to the instantiated sphere Mesh, correct? love how this is resolved so cleanly 👍

File: examples/component-example.py [hunk 11]
```diff
@@ -0,0 +1,122 @@
...
+            "Group",
+            self.state,
+            h(
+                Track,
+                {
+                    "scale": [0.5, 0.5, 10],
+                },
+            ),
+            h(Scrubber, {}),
+        )
```
Comment (Korijn) [inline examples/component-example.py:91]: now this is starting to look really cool here :)

File: collagraph/__init__.py (hunk 6 above)
Comment (Korijn) [reply collagraph/__init__.py:198]: Just the state should be sufficient.

File: collagraph/__init__.py (hunk 1 above)
Comment (Korijn) [reply collagraph/__init__.py:524]: depending on how critical some of these lifecycle hooks are, you could consider just leaving them out

File: collagraph/__init__.py (hunk 6 above)
Comment (Korijn) [reply collagraph/__init__.py:198]: If the props change, that is always caused somewhere higher up the component tree by a state change.

File: collagraph/component.py [hunk 12]
```diff
@@ -0,0 +1,79 @@
...
+        """Called after the component has been mounted.
+
+        A component is considered mounted after:
+
+        * All of its child components have been mounted.
+        * Its own DOM tree has been created and inserted into the parent container.
+        """
+        pass
+
+    def before_update(self):
```
[the remark this answers is not in the document]
Comment (berendkleinhaneveld) [reply collagraph/component.py:53]: Yes, let's do that. Should hopefully make things less complicated as well.

File: collagraph/component.py [hunk 13]
```diff
@@ -0,0 +1,79 @@
...
+            self._state[name] = value
+
+    def __getattribute__(self, name):
+        if name.startswith("_") or name in ["state", "props"]:
+            return super().__getattribute__(name)
+        if name in self._state:
+            return self._state[name]
+        if name in self._props:
+            return self._props[name]
+        return super().__getattribute__(name)
```
[the remark this answers is not in the document]
Comment (berendkleinhaneveld) [reply collagraph/component.py:37]: Hmmm, I guess you're right: explicit is better. I'll update this.

File: collagraph/__init__.py (hunk 6 above)
Comment (berendkleinhaneveld) [reply collagraph/__init__.py:198]: Yes, exactly. Should be sufficient like this!

File: examples/component-example.py (hunk 10 above)
Comment (berendkleinhaneveld) [reply examples/component-example.py:36]: Hehe, well, it still feels kind of hacky... I do think we'll want to implement something like Vues' `$el` and `$refs` functionality, but not for this PR yet ;)

File: collagraph/component.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (Korijn) [reply collagraph/component.py:53]: I guess "async rendering" here refers to fiber based rendering, just like you are doing here

File: examples/component-example.py (hunk 10 above)
Comment (Korijn) [reply examples/component-example.py:36]: In a way, they would just do the same kind of thing but just in response to a different event (mounted). So, hacky, maybe. I would say it's just less convenient. :)

File: collagraph/__init__.py (hunk 1 above)
Comment (berendkleinhaneveld) [reply collagraph/__init__.py]: Marking this as resolved.

File: examples/component-example.py (hunk 9 above)
Comment (berendkleinhaneveld) [reply examples/component-example.py]: No more `__setattr__` and `__getattribute__` tricks!

File: collagraph/__init__.py [hunk 14]
```diff
@@ -170,6 +181,31 @@ def perform_unit_of_work(self, fiber: Fiber) -> Optional[Fiber]:
...
+        if not component:
+            component = fiber.type(fiber.props)
+
+        # Attach the component instance to the fiber
+        fiber.component = component
+        fiber.component_watcher = watch(
+            # FIXME: sometimes fiber.component is None
+            # Seems to happen after a while after generating lots of changes.
+            # Could be that it is only when updates are coming in faster than
+            # they are rendered...
```
Comment (Korijn) [inline collagraph/__init__.py:195]: is this still an issue? can you reproduce it?

File: collagraph/__init__.py [hunk 15]
```diff
@@ -332,6 +369,39 @@ def commit_root(self):
             work = self._work.get()
             self.commit_work(work)
 
+        component_hooks = SimpleQueue()
+        component_hooks.put((self._wip_root, True))
```
Comment (Korijn) [inline collagraph/__init__.py:373]: can you add a comment to explain what the two elements of the tuples represent?

File: collagraph/__init__.py [hunk 16]
```diff
@@ -408,9 +503,10 @@ def is_equivalent_event_handler(val, other, key):
                 continue
 
             event_type = key_to_event(name)
-            self.renderer.remove_event_listener(dom, event_type, val)
+            events_to_remove[event_type] = val
```
Comment (Korijn) [inline collagraph/__init__.py:506]: just wondering if it's possible that there are multiple events to remove for a single event type?

File: collagraph/__init__.py [hunk 17]
```diff
@@ -439,7 +537,32 @@ def is_equivalent_event_handler(val, other, key):
                 continue
 
             event_type = key_to_event(name)
-            self.renderer.add_event_listener(dom, event_type, next_props[name])
+            events_to_add[event_type] = next_props[name]
```
Comment (Korijn) [inline collagraph/__init__.py:540]: same question as for events_to_remove

File: collagraph/component.py [hunk 18]
```diff
@@ -0,0 +1,41 @@
+from abc import ABCMeta, abstractmethod
+
+from observ import reactive, readonly
+
+
+class Component(metaclass=ABCMeta):
+    """Abstract base class for components"""
+
+    def __init__(self, props=None):
+        self.props = readonly(props or {})
```
Comment (Korijn) [inline collagraph/component.py:10]: An empty dict also evaluates to False, not sure if that might lead to surprising effects

File: collagraph/component.py [hunk 19]
```diff
@@ -0,0 +1,41 @@
...
+        self.state = reactive({})
+
+    def mounted(self):
+        """Called after the component has been mounted.
+
+        A component is considered mounted after:
```
Comment (Korijn) [inline collagraph/component.py:16]: You might want to be very clear here that this means the hook `is called after:`

File: examples/component-example.py [hunk 20]
```diff
@@ -0,0 +1,90 @@
...
+    materials = {
+        "default": gfx.MeshPhongMaterial(color=[1.0, 0.5, 0.0]),
+        "pressed": gfx.MeshPhongMaterial(color=[0.0, 0.5, 0.0]),
+    }
+
+    def __init__(self, props):
+        super().__init__(props)
+        self.state["pressed"] = False
+        self.state["scale"] = self.props.get("scale", [0.85] * 3)
+        self.mesh = None
```
Comment (Korijn) [inline examples/component-example.py:19]: Just wanted to say that this looks great, it's very clear now that `self.mesh` is not a prop, not state, and not reactive.

File: examples/component-example.py [hunk 21]
```diff
@@ -0,0 +1,90 @@
...
+
+    element = h(NumberPad, {})
+    container = gfx.Scene()
+
+    def animate():
+        controls.update_camera(camera)
+        renderer.render(container, camera)
+
+    gui.render(element, container, callback=lambda: canvas.request_draw(animate))
+    canvas.request_draw(animate)
```
Comment (Korijn) [inline examples/component-example.py:89]: is this line necessary? can't you leave it up to collagraph to do the first call?

File: collagraph/__init__.py (hunk 16 above)
Comment (berendkleinhaneveld) [reply collagraph/__init__.py:506]: Right, I guess we need to replicate the same data structure: dict with sets as values.

File: collagraph/component.py (hunk 18 above)
Comment (berendkleinhaneveld) [reply collagraph/component.py:10]: Ah, right: in order to preserve reactivity... I'll just check for `is None` instead.

File: examples/component-example.py (hunk 20 above)
Comment (berendkleinhaneveld) [reply examples/component-example.py:19]: Yes! I also believe that it is unused currently, so I'll remove it anyways ;)

File: collagraph/__init__.py (hunk 16 above)
Comment (berendkleinhaneveld) [reply collagraph/__init__.py:506]: Actually: the question is: do we want to support the following?

```python
class Comp(Component):
    def call_a(self, event):
        pass

    def call_b(self, event):
        pass

    def render(self):
        return h("x", {"on_event": {self.call_a, self.call_b}})
```

I think that we don't want that (also in the case that we add JSX-like syntax to specify components), so I think the code is fine as it is.

File: collagraph/component.py (hunk 19 above)
Comment (berendkleinhaneveld) [reply collagraph/component.py:16]: The doc string already states this (line 14), so I'm not sure what you're getting at. Does that suffice?

File: collagraph/__init__.py (hunk 14 above)
Comment (berendkleinhaneveld) [reply collagraph/__init__.py:195]: Probably not with the current example, but I'm sure it still exists. I'll make an issue about this such that we don't forget about it.

Comment (berendkleinhaneveld) [reply collagraph/__init__.py:195]: I've created #17 to keep track of this.

File: collagraph/__init__.py (hunk 16 above)
Comment (Korijn) [reply collagraph/__init__.py:515]: Good call!

**Comment by Korijn** [comment]
LGTM!

**Pull request merged**